kernel/task/src/lib.rs

```diff
@@ -953,70 +953,32 @@ fn task_switch_inner(
-    /// and then retrieved from here right after the context switch ends.
-    /// It is stored in a CPU-local variable because it's only related to
-    /// a task switching operation on a particular CPU.
-    pub(crate) static TASK_SWITCH_PREEMPTION_GUARD: CpuLocal<TaskSwitchPreemptionGuard> =
-        CpuLocal::new(PerCpuField::TaskSwitchPreemptionGuard);
-
-    /// Data that should be dropped after switching away from a task that has exited.
-    ///
-    /// Currently, this contains the previous Task's `TaskRef` removed from its TLS area;
-    /// it is stored in a CPU-local variable because it's only related to
-    /// a task switching operation on a particular CPU.
-    pub(crate) static DROP_AFTER_TASK_SWITCH: CpuLocal<DropAfterTaskSwitch> =
-        CpuLocal::new(PerCpuField::DropAfterTaskSwitch);
-
-    /// A type wrapper used to hold a CPU-local `PreemptionGuard` 
-    /// on the current CPU during a task switch operation.
-    #[derive(Default)]
-    pub struct TaskSwitchPreemptionGuard(pub(crate) Option<PreemptionGuard>);
-    impl TaskSwitchPreemptionGuard {
-        pub const fn new() -> Self { Self(None) }
-    }
-    // SAFETY: The `TaskSwitchPreemptionGuard` type corresponds to a field in `PerCpuData`
-    //         with the offset specified by `PerCpuField::TaskSwitchPreemptionGuard`.
-    unsafe impl CpuLocalField for TaskSwitchPreemptionGuard {
-        const FIELD: PerCpuField = PerCpuField::TaskSwitchPreemptionGuard;
-    }
-
-    /// A type wrapper used to hold CPU-local data that should be dropped
-    /// after switching away from a task that has exited.
-    #[derive(Default)]
-    pub struct DropAfterTaskSwitch(pub(crate) Option<super::TaskRef>);
-    impl DropAfterTaskSwitch {
-        pub const fn new() -> Self { Self(None) }
-    }
-    // SAFETY: The `DropAfterTaskSwitch` type corresponds to a field in `PerCpuData`
-    //         with the offset specified by `PerCpuField::DropAfterTaskSwitch`.
-    unsafe impl CpuLocalField for DropAfterTaskSwitch {
-        const FIELD: PerCpuField = PerCpuField::DropAfterTaskSwitch;
-    }
-}
-
+/// The preemption guard that was used for safe task switching on each CPU.
+///
+/// The `PreemptionGuard` is stored here right before a context switch begins
+/// and then retrieved from here right after the context switch ends.
+/// It is stored in a CPU-local variable because it's only related to
+/// a task switching operation on a particular CPU.
+#[cls::cpu_local(16, stores_guard = PreemptionGuard)]
+pub(crate) static TASK_SWITCH_PREEMPTION_GUARD: Option<PreemptionGuard> = None;
+
+/// Data that should be dropped after switching away from a task that has exited.
+///
+/// Currently, this contains the previous Task's `TaskRef` removed from its TLS area;
+/// it is stored in a CPU-local variable because it's only related to
+/// a task switching operation on a particular CPU.
+#[cls::cpu_local(24)]
+pub(crate) static DROP_AFTER_TASK_SWITCH: Option<TaskRef> = None;
```

Comment:
nit: since we moved these out of an inner module, they don't need to have any visibility modifier like `pub(crate)`